app/controllers/api/v1/study_files_controller.rb

```diff
@@ -291,60 +301,96 @@ def perform_update(study_file)
-            @study.update(default_options: @study.default_options.merge(expression_label: safe_file_params[:y_axis_label]))
-          end
+        if ['Expression Matrix', 'MM Coordinate Matrix'].include?(safe_file_params[:file_type]) && !safe_file_params[:y_axis_label].blank?
+          # if user is supplying an expression axis label, update default options hash
+          options = @study.default_options.dup
+          options.merge!(expression_label: safe_file_params[:y_axis_label])
+          @study.update(default_options: options)
+        end
 
-          if parse_on_upload &&
-            study_file.parseable? &&
-            !study_file.parsing? &&
-            study_file.parse_status == 'unparsed' &&
-            study_file.status == 'uploaded' &&
-            safe_file_params[:upload].present?
-            FileParseService.run_parse_job(@study_file, @study, current_api_user)
-          end
+        if safe_file_params[:upload].present? && !is_chunked
+          complete_upload_process(study_file, parse_on_upload)
+        end
+      end
 
-          return true
+
+      # handles adding an additional chunk to a file upload.  Note that create/update must be
+      # called first with the first chunk of the file, in order to initialize the process
+      # if this is the last chunk, it will handle sending to firecloud and launching a
+      # parse job if requested
+      def chunk
+        safe_file_params = study_file_params
+        upload = safe_file_params[:upload]
+        content_range = RequestUtils.parse_content_range_header(request.headers)
+        if content_range.present?
+          if content_range[:first_byte] == 0
+            render json: {errors: 'create/update should be used for uploading the first chunk'}, status: :unprocessable_entity
+          end
+          if content_range[:first_byte] != @study_file.upload_file_size
+            render json: {errors: "Incorrect chunk sent: expected bytes starting with #{@study_file.upload_file_size}, received #{content_range[:first_byte]}" }, status: :unprocessable_entity
+          end
         else
-          return false
+          render json: {errors: 'Missing Content-Range header'}, status: :unprocessable_entity
+        end
+
+        File.open(@study_file.upload.path, "ab") do |f|
+          f.write upload.read
+        end
+
+        # Update the upload_file_size attribute
+        @study_file.upload_file_size = @study_file.upload_file_size.nil? ? upload.size : @study_file.upload_file_size + upload.size
+        @study_file.save!
+
+        if @study_file.upload_file_size >= content_range[:total_size]
+          # this was the last chunk
+          complete_upload_process(@study_file, safe_file_params[:parse_on_upload])
         end
       end
 
+      def complete_upload_process(study_file, parse_on_upload)
+        @study.delay.send_to_firecloud(study_file) # send data to FireCloud if upload was performed
```

Comment:
To inform potential optimizations, could you wrap this call with timers and log that to Mixpanel (along with file type, size)?